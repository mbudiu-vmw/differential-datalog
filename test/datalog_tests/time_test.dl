import time

input relation TITest(t: Time)
output relation TTest(s: string, t: Result<Time, string>)

TTest(t.to_string(), Ok{t}) :- TITest(t).
TTest("try_from_hms(8'd10, 8'd10, 8'd10)", try_from_hms(8'd10, 8'd10, 8'd10)).
TTest("try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10)", try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10)).
TTest("try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10)", try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10)).
TTest("try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10)", try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10)).
TTest("try_from_hms(8'd23, 8'd0, 8'd0)", try_from_hms(8'd23, 8'd0, 8'd0)).
TTest("try_from_hms(8'd24, 8'd0, 8'd0)", try_from_hms(8'd24, 8'd0, 8'd0)).
TTest("try_from_hms(8'd0, 8'd60, 8'd0)", try_from_hms(8'd0, 8'd60, 8'd0)).
TTest("Ok(midnight())", Ok{midnight()}: Result<Time, string>).
TTest("time_parse(\"10:10\", \"%T\")", time_parse("10:10", "%T")).
TTest("time_parse(\"10:10:10\", \"%T\")", time_parse("10:10:10", "%T")).
TTest("time_parse(\"10:10:10.102030405\", \"%T.%f\")", time_parse("10:10:10.102030405", "%T.%f")).
TTest("string2time(to_string(result_unwrap_or_default(try_from_hms(8'd10, 8'd10, 8'd10))))",
       try_from_hms(8'd10, 8'd10, 8'd10).unwrap_or_default().to_string().to_time()).
TTest("string2time(to_string(result_unwrap_or_default(try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10))))",
       string2time(to_string(result_unwrap_or_default(try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10))))).
TTest("string2time(to_string(result_unwrap_or_default(try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10))))",
       string2time(to_string(result_unwrap_or_default(try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10))))).
TTest("string2time(to_string(result_unwrap_or_default(try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10))))",
       string2time(to_string(result_unwrap_or_default(try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10))))).
TTest("time_parse(time_format(${someTime()}, ${someTimeFormat()}), ${someTimeFormat()})",
       time_parse(time_format(someTime(), someTimeFormat()), someTimeFormat())).

function someTime(): Time {
   result_unwrap_or_default( try_from_hms_nano(10, 11, 12, 103104105) )
}
function someTimeFormat(): string {
   "%H::%M::%S"
}

output relation Extract(s: string, v: bit<32>)
Extract("hour(${someTime()})", hour(someTime()) as bit<32>).
Extract("minute(${someTime()})", minute(someTime()) as bit<32>).
Extract("second(${someTime()})", second(someTime()) as bit<32>).
Extract("millisecond(${someTime()})", millisecond(someTime()) as bit<32>).
Extract("microsecond(${someTime()})", microsecond(someTime()) as bit<32>).
Extract("nanosecond(${someTime()})", nanosecond(someTime()) as bit<32>).

///////////////////////////////////////

function someDate(): Date {
   result_unwrap_or_default( try_from_ymd(32'sd2020, 8'd04, 8'd14) )
}
function someDateFormat(): string {
   "%Y-%m-%dT"
}

input relation DITest(d: Date)
output relation DTest(s: string, t: Result<Date, string>)

DTest(to_string(d), Ok{d}) :- DITest(d).
DTest("try_from_ymd(32'sd2020, 8'd04, 8'd14)", try_from_ymd(32'sd2020, 8'd04, 8'd14)).
DTest("try_from_ymd(32'sd2020, 8'd13, 8'd14)", try_from_ymd(32'sd2020, 8'd13, 8'd14)).
DTest("try_from_ymd(32'sd2020, 8'd0, 8'd14)", try_from_ymd(32'sd2020, 8'd0, 8'd14)).
DTest("try_from_yo(32'sd2020, 16'd100)", try_from_yo(32'sd2020, 16'd100)).
DTest("try_from_yo(32'sd2020, 16'd367)", try_from_yo(32'sd2020, 16'd367)).
DTest("try_from_iso_ywd(32'sd2020, 8'd4, Monday)", try_from_iso_ywd(32'sd2020, 8'd4, Monday)).
DTest("date_parse(\"2020/04/14\", \"%Y/%m/%d\")", date_parse("2020/04/14", "%Y/%m/%d")).
DTest("string2date(to_string(someDate()))",
       string2date(to_string(someDate()))).
DTest("Ok{next_day(${someDate()})}", Ok{next_day(someDate())}).
DTest("Ok{previous_day(${someDate()})}", Ok{previous_day(someDate())}).
DTest("Ok{from_julian_day(2000000)}", Ok{from_julian_day(2000000)}).
DTest("date_parse(date_format(${someDate()}, ${someDateFormat()}), ${someDateFormat()})",
       date_parse(date_format(someDate(), someDateFormat()), someDateFormat())).

Extract("year(${someDate()})", year(someDate()) as bit<32>).
Extract("month(${someDate()})", month(someDate()) as bit<32>).
Extract("day(${someDate()})", day(someDate()) as bit<32>).
Extract("ordinal(${someDate()})", ordinal(someDate()) as bit<32>).
Extract("week(${someDate()})", week(someDate()) as bit<32>).

//////////////////////////////////////////////

function someDateTime(): DateTime {
   DateTime{.date = someDate(), .time = someTime()}
}
function someDateTimeFormat(): string {
   "%Y-%m-%dT%H::%M"
}

input relation DTITest(d: DateTime)
output relation DTTest(s: string, t: Result<DateTime, string>)

function get_datetime(r: Result<TzDateTime, string>): Result<DateTime, string> {
     match (r) {
         Ok{dt} -> Ok{utc_datetime(dt)},
         Err{e} -> Err{e}
     }
}

DTTest(to_string(d), Ok{d}) :- DTITest(d).
DTTest("string2datetime(\"2020-10-10T10:20:30\")", string2datetime("2020-10-10T10:20:30")).
DTTest("string2datetime(to_string(${someDateTime()}))",
        string2datetime(to_string(someDateTime()))).
DTTest("datetime_parse(\"2020-10-10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")",
        datetime_parse("2020-10-10T10:20:30", "%Y-%m-%dT%H:%M:%S")).
DTTest("datetime_parse(\"2020/10/10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")",
        datetime_parse("2020/10/10T10:20:30", "%Y-%m-%dT%H:%M:%S")).
DTTest("datetime_parse(datetime_format(${someDateTime()}, ${someDateTimeFormat()}), ${someDateTimeFormat()})",
        datetime_parse(datetime_format(someDateTime(), someDateTimeFormat()), someDateTimeFormat())).
DTTest("datetime_from_unix_timestamp(10000000)", Ok{datetime_from_unix_timestamp(10000000)}).
DTTest("datetime_parse(\"Wed Mar 05 07:45:30 MST 2025\", \"%a %b %d %H:%M:%S %Z %Y\")",
        datetime_parse("Wed Mar 05 07:45:30 MST 2025", "%a %b %d %H:%M:%S %Z %Y")).
DTTest("get_datetime(tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 GMT\"))",
        get_datetime(tz_datetime_parse_from_rfc2822("Tue, 1 Jul 2003 10:52:37 GMT"))).

//////////////////////////////////////////////////

input relation DTZITest(d: TzDateTime)
output relation DTZTest(s: string, t: Result<TzDateTime, string>)

function utc(r: Result<DateTime, string>): Result<TzDateTime, string> {
     match (r) {
         Ok{dt} -> Ok{utc_timezone(dt)},
         Err{e} -> Err{e}
     }
}

DTZTest(to_string(d), Ok{d}) :- DTZITest(d).
DTZTest("utc(string2datetime(\"2020-10-10T10:20:30\"))", utc(string2datetime("2020-10-10T10:20:30"))).
DTZTest("tz_datetime_parse(\"2020-10-10T10:20:30 +02:00\", \"%Y-%m-%dT%H:%M:%S %z\")",
        tz_datetime_parse("2020-10-10T10:20:30 +02:00", "%Y-%m-%dT%H:%M:%S %z")).
DTZTest("tz_datetime_parse(\"2020/10/10T10:20:30 -02:00\", \"%Y/%m/%dT%H:%M:%S %z\")",
        tz_datetime_parse("2020/10/10T10:20:30 -03:00", "%Y/%m/%dT%H:%M:%S %z")).
DTZTest("tz_datetime_parse_from_rfc3339(\"2020-10-10T10:20:30-03:00\")",
        tz_datetime_parse_from_rfc3339("2020-10-10T10:20:30-03:00")).
DTZTest("tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 +0200\")",
        tz_datetime_parse_from_rfc2822("Tue, 1 Jul 2003 10:52:37 +0200")).
DTZTest("tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 GMT\")",
        tz_datetime_parse_from_rfc2822("Tue, 1 Jul 2003 10:52:37 GMT")).
// The following actually gives an error, since the %Z format cannot infer a timezone reliably
// So this cannot extract a date with a timezone, only a DateTime
DTZTest("tz_datetime_parse(\"Wed Mar 05 07:45:30 MST 2025\", \"%a %b %d %H:%M:%S %Z %Y\")",
         tz_datetime_parse("Wed Mar 05 07:45:30 MST 2025", "%a %b %d %H:%M:%S %Z %Y")).
