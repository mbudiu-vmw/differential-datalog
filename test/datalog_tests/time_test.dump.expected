DTTest:
DTTest{.s = "datetime_from_unix_timestamp(10000000)", .t = ddlog_std::Ok{.res = time::DateTime{.date = "1970-04-26", .time = "17:46:40.000000000"}}}: +1
DTTest{.s = "datetime_parse(\"2020-10-10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-10-10", .time = "10:20:30.000000000"}}}: +1
DTTest{.s = "datetime_parse(\"2020/10/10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")", .t = ddlog_std::Err{.err = "input contains invalid characters"}}: +1
DTTest{.s = "datetime_parse(\"Wed Mar 05 07:45:30 MST 2025\", \"%a %b %d %H:%M:%S %Z %Y\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2025-03-05", .time = "07:45:30.000000000"}}}: +1
DTTest{.s = "datetime_parse(datetime_format(2020-04-14T10:11:12, %Y-%m-%dT%H::%M), %Y-%m-%dT%H::%M)", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-04-14", .time = "10:11:00.000000000"}}}: +1
DTTest{.s = "get_datetime(tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 GMT\"))", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2003-07-01", .time = "10:52:37.000000000"}}}: +1
DTTest{.s = "string2datetime(\"2020-10-10T10:20:30\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-10-10", .time = "10:20:30.000000000"}}}: +1
DTTest{.s = "string2datetime(to_string(2020-04-14T10:11:12))", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-04-14", .time = "10:11:12.000000000"}}}: +1
DTZTest:
DTZTest{.s = "tz_datetime_parse(\"2020-10-10T10:20:30 +02:00\", \"%Y-%m-%dT%H:%M:%S %z\")", .t = ddlog_std::Ok{.res = "2020-10-10T10:20:30+02:00"}}: +1
DTZTest{.s = "tz_datetime_parse(\"2020/10/10T10:20:30 -02:00\", \"%Y/%m/%dT%H:%M:%S %z\")", .t = ddlog_std::Ok{.res = "2020-10-10T10:20:30-03:00"}}: +1
DTZTest{.s = "tz_datetime_parse(\"Wed Mar 05 07:45:30 MST 2025\", \"%a %b %d %H:%M:%S %Z %Y\")", .t = ddlog_std::Err{.err = "input is not enough for unique date and time"}}: +1
DTZTest{.s = "tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 +0200\")", .t = ddlog_std::Ok{.res = "2003-07-01T10:52:37+02:00"}}: +1
DTZTest{.s = "tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 GMT\")", .t = ddlog_std::Ok{.res = "2003-07-01T10:52:37+00:00"}}: +1
DTZTest{.s = "tz_datetime_parse_from_rfc3339(\"2020-10-10T10:20:30-03:00\")", .t = ddlog_std::Ok{.res = "2020-10-10T10:20:30-03:00"}}: +1
DTZTest{.s = "utc(string2datetime(\"2020-10-10T10:20:30\"))", .t = ddlog_std::Ok{.res = "2020-10-10T10:20:30+00:00"}}: +1
DTest:
DTest{.s = "Ok{from_julian_day(2000000)}", .t = ddlog_std::Ok{.res = "0763-09-18"}}: +1
DTest{.s = "Ok{next_day(2020-04-14)}", .t = ddlog_std::Ok{.res = "2020-04-15"}}: +1
DTest{.s = "Ok{previous_day(2020-04-14)}", .t = ddlog_std::Ok{.res = "2020-04-13"}}: +1
DTest{.s = "date_parse(\"2020/04/14\", \"%Y/%m/%d\")", .t = ddlog_std::Err{.err = "input contains invalid characters"}}: +1
DTest{.s = "date_parse(date_format(2020-04-14, %Y-%m-%dT), %Y-%m-%dT)", .t = ddlog_std::Err{.err = "trailing input"}}: +1
DTest{.s = "string2date(to_string(someDate()))", .t = ddlog_std::Ok{.res = "2020-04-14"}}: +1
DTest{.s = "try_from_iso_ywd(32'sd2020, 8'd4, Monday)", .t = ddlog_std::Ok{.res = "2020-01-20"}}: +1
DTest{.s = "try_from_ymd(32'sd2020, 8'd0, 8'd14)", .t = ddlog_std::Err{.err = "Invalid date"}}: +1
DTest{.s = "try_from_ymd(32'sd2020, 8'd04, 8'd14)", .t = ddlog_std::Ok{.res = "2020-04-14"}}: +1
DTest{.s = "try_from_ymd(32'sd2020, 8'd13, 8'd14)", .t = ddlog_std::Err{.err = "Invalid date"}}: +1
DTest{.s = "try_from_yo(32'sd2020, 16'd100)", .t = ddlog_std::Ok{.res = "2020-04-09"}}: +1
DTest{.s = "try_from_yo(32'sd2020, 16'd367)", .t = ddlog_std::Err{.err = "Invalid date"}}: +1
Extract:
Extract{.s = "day(2020-04-14)", .v = 14}: +1
Extract{.s = "hour(10:11:12.103104105)", .v = 10}: +1
Extract{.s = "microsecond(10:11:12.103104105)", .v = 103104}: +1
Extract{.s = "millisecond(10:11:12.103104105)", .v = 103}: +1
Extract{.s = "minute(10:11:12.103104105)", .v = 11}: +1
Extract{.s = "month(2020-04-14)", .v = 4}: +1
Extract{.s = "nanosecond(10:11:12.103104105)", .v = 103104105}: +1
Extract{.s = "ordinal(2020-04-14)", .v = 105}: +1
Extract{.s = "second(10:11:12.103104105)", .v = 12}: +1
Extract{.s = "week(2020-04-14)", .v = 16}: +1
Extract{.s = "year(2020-04-14)", .v = 2020}: +1
TTest:
TTest{.s = "Ok(midnight())", .t = ddlog_std::Ok{.res = "00:00:00.000000000"}}: +1
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms(8'd10, 8'd10, 8'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}: +1
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000010000"}}: +1
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.010000000"}}: +1
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000000010"}}: +1
TTest{.s = "time_parse(\"10:10\", \"%T\")", .t = ddlog_std::Err{.err = "premature end of input"}}: +1
TTest{.s = "time_parse(\"10:10:10\", \"%T\")", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}: +1
TTest{.s = "time_parse(\"10:10:10.102030405\", \"%T.%f\")", .t = ddlog_std::Ok{.res = "10:10:10.102030405"}}: +1
TTest{.s = "time_parse(time_format(10:11:12.103104105, %H::%M::%S), %H::%M::%S)", .t = ddlog_std::Ok{.res = "10:11:12.000000000"}}: +1
TTest{.s = "try_from_hms(8'd0, 8'd60, 8'd0)", .t = ddlog_std::Err{.err = "illegal time value"}}: +1
TTest{.s = "try_from_hms(8'd10, 8'd10, 8'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}: +1
TTest{.s = "try_from_hms(8'd23, 8'd0, 8'd0)", .t = ddlog_std::Ok{.res = "23:00:00.000000000"}}: +1
TTest{.s = "try_from_hms(8'd24, 8'd0, 8'd0)", .t = ddlog_std::Err{.err = "illegal time value"}}: +1
TTest{.s = "try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000010000"}}: +1
TTest{.s = "try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10)", .t = ddlog_std::Ok{.res = "10:10:10.010000000"}}: +1
TTest{.s = "try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000000010"}}: +1
TTest{.s = "10:11:12.000000000", .t = ddlog_std::Ok{.res = "10:11:12.000000000"}}
TTest{.s = "Ok(midnight())", .t = ddlog_std::Ok{.res = "00:00:00.000000000"}}
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms(8'd10, 8'd10, 8'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000010000"}}
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.010000000"}}
TTest{.s = "string2time(to_string(result_unwrap_or_default(try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10))))", .t = ddlog_std::Ok{.res = "10:10:10.000000010"}}
TTest{.s = "time_parse(\"10:10\", \"%T\")", .t = ddlog_std::Err{.err = "premature end of input"}}
TTest{.s = "time_parse(\"10:10:10\", \"%T\")", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}
TTest{.s = "time_parse(\"10:10:10.102030405\", \"%T.%f\")", .t = ddlog_std::Ok{.res = "10:10:10.102030405"}}
TTest{.s = "time_parse(time_format(10:11:12.103104105, %H::%M::%S), %H::%M::%S)", .t = ddlog_std::Ok{.res = "10:11:12.000000000"}}
TTest{.s = "try_from_hms(8'd0, 8'd60, 8'd0)", .t = ddlog_std::Err{.err = "illegal time value"}}
TTest{.s = "try_from_hms(8'd10, 8'd10, 8'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000000000"}}
TTest{.s = "try_from_hms(8'd23, 8'd0, 8'd0)", .t = ddlog_std::Ok{.res = "23:00:00.000000000"}}
TTest{.s = "try_from_hms(8'd24, 8'd0, 8'd0)", .t = ddlog_std::Err{.err = "illegal time value"}}
TTest{.s = "try_from_hms_micro(8'd10, 8'd10, 8'd10, 32'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000010000"}}
TTest{.s = "try_from_hms_milli(8'd10, 8'd10, 8'd10, 16'd10)", .t = ddlog_std::Ok{.res = "10:10:10.010000000"}}
TTest{.s = "try_from_hms_nano(8'd10, 8'd10, 8'd10, 32'd10)", .t = ddlog_std::Ok{.res = "10:10:10.000000010"}}
Extract{.s = "day(2020-04-14)", .v = 14}
Extract{.s = "hour(10:11:12.103104105)", .v = 10}
Extract{.s = "microsecond(10:11:12.103104105)", .v = 103104}
Extract{.s = "millisecond(10:11:12.103104105)", .v = 103}
Extract{.s = "minute(10:11:12.103104105)", .v = 11}
Extract{.s = "month(2020-04-14)", .v = 4}
Extract{.s = "nanosecond(10:11:12.103104105)", .v = 103104105}
Extract{.s = "ordinal(2020-04-14)", .v = 105}
Extract{.s = "second(10:11:12.103104105)", .v = 12}
Extract{.s = "week(2020-04-14)", .v = 16}
Extract{.s = "year(2020-04-14)", .v = 2020}
DTest{.s = "2010-10-15", .t = ddlog_std::Ok{.res = "2010-10-15"}}
DTest{.s = "Ok{from_julian_day(2000000)}", .t = ddlog_std::Ok{.res = "0763-09-18"}}
DTest{.s = "Ok{next_day(2020-04-14)}", .t = ddlog_std::Ok{.res = "2020-04-15"}}
DTest{.s = "Ok{previous_day(2020-04-14)}", .t = ddlog_std::Ok{.res = "2020-04-13"}}
DTest{.s = "date_parse(\"2020/04/14\", \"%Y/%m/%d\")", .t = ddlog_std::Err{.err = "input contains invalid characters"}}
DTest{.s = "date_parse(date_format(2020-04-14, %Y-%m-%dT), %Y-%m-%dT)", .t = ddlog_std::Err{.err = "trailing input"}}
DTest{.s = "string2date(to_string(someDate()))", .t = ddlog_std::Ok{.res = "2020-04-14"}}
DTest{.s = "try_from_iso_ywd(32'sd2020, 8'd4, Monday)", .t = ddlog_std::Ok{.res = "2020-01-20"}}
DTest{.s = "try_from_ymd(32'sd2020, 8'd0, 8'd14)", .t = ddlog_std::Err{.err = "Invalid date"}}
DTest{.s = "try_from_ymd(32'sd2020, 8'd04, 8'd14)", .t = ddlog_std::Ok{.res = "2020-04-14"}}
DTest{.s = "try_from_ymd(32'sd2020, 8'd13, 8'd14)", .t = ddlog_std::Err{.err = "Invalid date"}}
DTest{.s = "try_from_yo(32'sd2020, 16'd100)", .t = ddlog_std::Ok{.res = "2020-04-09"}}
DTest{.s = "try_from_yo(32'sd2020, 16'd367)", .t = ddlog_std::Err{.err = "Invalid date"}}
DTTest{.s = "2010-10-15T10:11:12", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2010-10-15", .time = "10:11:12.000000000"}}}
DTTest{.s = "datetime_from_unix_timestamp(10000000)", .t = ddlog_std::Ok{.res = time::DateTime{.date = "1970-04-26", .time = "17:46:40.000000000"}}}
DTTest{.s = "datetime_parse(\"2020-10-10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-10-10", .time = "10:20:30.000000000"}}}
DTTest{.s = "datetime_parse(\"2020/10/10T10:20:30\", \"%Y-%m-%dT%H:%M:%S\")", .t = ddlog_std::Err{.err = "input contains invalid characters"}}
DTTest{.s = "datetime_parse(\"Wed Mar 05 07:45:30 MST 2025\", \"%a %b %d %H:%M:%S %Z %Y\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2025-03-05", .time = "07:45:30.000000000"}}}
DTTest{.s = "datetime_parse(datetime_format(2020-04-14T10:11:12, %Y-%m-%dT%H::%M), %Y-%m-%dT%H::%M)", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-04-14", .time = "10:11:00.000000000"}}}
DTTest{.s = "get_datetime(tz_datetime_parse_from_rfc2822(\"Tue, 1 Jul 2003 10:52:37 GMT\"))", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2003-07-01", .time = "10:52:37.000000000"}}}
DTTest{.s = "string2datetime(\"2020-10-10T10:20:30\")", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-10-10", .time = "10:20:30.000000000"}}}
DTTest{.s = "string2datetime(to_string(2020-04-14T10:11:12))", .t = ddlog_std::Ok{.res = time::DateTime{.date = "2020-04-14", .time = "10:11:12.000000000"}}}
